### ISSUE TITLE
Bug: the Mudlet client replaces all instances of '<' and '>' with '&lt' and '&gt'

### ISSUE THREAD

**Issue by TehomCD** [evennia/evennia#1058]
#### Brief summary of issue / Description of requested feature:

After upgrading to the latest version of Evennia, a player on mudlet reported seeing all output that had those symbols transformed. Downloading mudlet and checking myself reproduced the issue.
#### Steps to reproduce the issue / Reasons for adding feature:
1. Download mudlet, connect, 'say < test >'
#### Error output / Expected result of feature

expected: 'You say: < test >', currently: 'You say: &lt test &gt'

**Comment by Griatch** [comment]
This is an issue with Mudlet reporting that it wants MXP replacement even though it apparently does not. MXP negotiation protocol dictates that once the client has told the server (Evennia, in this case) that it supports MXP, the server must send all literal < > as their &lt; &gt; html counterparts, since <> tags are reserved for use in MXP. 
As far as we can discern, Mudlet tells Evennia to switch to MXP mode and then doesn't actually handle the &lt; &gt;.

That said, I don't know if this comes from some particular setting or version of mudlet, for I cannot replicate it here. More info would be needed to see if Evennia could assume some other default for Mudlet in particular. 

Anyway, using options/save MXP=False` will manually turn off MXP mode.

**Comment by BlauFeuer** [comment]
At first glance, this looks like a duplicate of #1020. 

It seems like MXP mode is on by default and has a different interpretation of the MXP standard in some versions of Mudlet (also on Github, here: https://github.com/Mudlet/Mudlet )

**Comment by TehomCD** [comment]
The version of Mudlet I'm seeing the issue on is 2.1. Here's an oddity - when I checked @option for a character on mudlet, MXP was already listed as False. But '@options/save MXP=False' still resolved the issue.

**Comment by Griatch** [comment]
The `@option/save` discrepancy seen by @TehomCD was a bug in the syncing to the option dict to the Server side. It's irrelevant to this Mudlet issue, but it was confusing and I've fixed it now. 

As far as I can tell when putting some printouts in the telnet MXP protocol, the correct process happens as far as Evennia is concerned: 

1. Evennia asks Mudlet if it is able to do MXP (IAC SB WILL MXP IAC SE)
2. Mudlet replies yes, I'm able to do MXP (IAC SB DO MXP IAC SE)
3. Evennia says, okay, let's switch to MXP mode then (IAC SB MXP IAC SE) 
4. Evennia switches to MXP mode, which means converting literal `<`, `>` and `&` to `&lt;` etc. (since < > are used in MXP tags). It henceforth assumes Mudlet knows how to handle these. 
5. Mudlet doesn't appear to switch to MXP mode though, but keeps printing the `&lt;` etc verbatim.

I have sent a question to the `#mudlet` IRC channel to ask if there is some known issue or special thing that must be done when sending to mudlet. But I don't know how active that channel actually is; Mudlet was last updated in 2013.

**Comment by BlauFeuer** [comment]
Mudlet is currently in [active development here](https://github.com/Mudlet/Mudlet), hence the reports of people testing [Mudlet 3.0.0 delta preview #4](http://forums.mudlet.org/viewtopic.php?f=5&t=4695) in #1020 Bug: Mudlet Isssues Displaying < and >

**Comment by Griatch** [comment]
The github page for mudlet 3.0 has no Issue tracker. I have repeated the question about this in the `#mudlet` irc channel daily for the past week. Others ask there too, but there doesn't seem to be any developer activity (or interest) in there. So at this point, I can only conclude Mudlet's MXP is broken since it doesn't switch over to parsing `&lt;` `&gt;` despite telling Evennia that it wants to do MXP.

**Comment by Griatch** [comment]
This is actually resolved with the latest update - the `<  > ` will now display correctly with MXP active. It is however replaced with another error: Now double quotes, like those used for the `say` command, are shown as `&amp;` instead in Mudlet.

**Comment by Griatch** [comment]
@BlauFeuer Could the way mudlet shows `&amp;` instead of double-quotes now be related to the latest changes to MXP output? Most of that was for the webclient but there were some changes to the portal/mxp.py file too.

**Comment by BlauFeuer** [comment]
Mudlet's latest release is 3.0.0 epsilon (preview #5). The contact email is vadim.peretokin@mudlet.org This is probably the older contributor, vperetokin@gmail.com They are listed as being on irc.freenote.net on the `#mudlet` channel, but perhaps the developers don't hang out there. Other places I see activity are Twitter: http://twitter.com/mudlet, Facebook: mudlet group and http://forums.mudlet.org/  On github,  `@SlySven` is the main active contributor, `@vadi2` not being as active anymore. `@dicene` has recently made contributions.

I have never actually used Mudlet, but it's likely that the `&quot;` being sent when Evennia's set for a session in MXP mode is being turned to `&amp;` as it begins to interpret the `&` in `&quot;` when sent something like `character says, "Wow!"` it is sent back as `You say, &quot;Wow!&quot;` to an MXP-enabled client.  

The change I intended to make was only to encode the double quote within the context of the MXP command, to turn something like this:  `say |lcMy "|ltIt's a "|le` into this: `You say, &quot;<SEND HREF="My &quot;">It's a &quot;</SEND>&quot;`

Maybe Mudlet is interpreting the protocol standard to only expect HTML-entitiy-like encodings within the html-like-tags of MXP.   In the context of sending to the webclient, all double quotes must be encoded into the `&quot;` HTML entity to be displayed correctly.  Perhaps Evennia's now being overly-aggressive in the scope of double quote encoding outside of the MXP tags.  Let's go look at the MXP standard again to make sure Evennia's doing it right.  As for Mudlet, it could be a bug that they need to address.

**Comment by BlauFeuer** [comment]
From what I read in [the standard here](http://www.zuggsoft.com/zmud/mxp.htm#Implementation%20Details), The text portion (in Evennia markup, the stuff between `|lt` and `|le`) does not require surrounding by quotes, so only needs quote escaping done within the context of the MXP command (in Evennia markup, the stuff between the `|lc` and the `|lt`) - with the exception of webclient where it must be done to display properly.

**Comment by BlauFeuer** [comment]
Assuming that interpretation is correct, Evennia (in Telnet clients) should interpret 

`say |lcMy "|ltIt's a "|le`

as

`You say, "<SEND HREF="My &quot;">It's a "</SEND>"`

and not as it currently does:

`You say, &quot;<SEND HREF="My &quot;">It's a &quot;</SEND>&quot;`

(In the case of web client, the double quote has to be encoded into an HTML entity.)

Someone else should look at this and try to interpret the standard, too.  It might very well be that the problem between Evennia and Mudlet is simply a difference of interpretation of the MXP standard and not a bug in either.

**Comment by Griatch** [comment]
My bad, I wrote `&amp;` twice in the comments above for some reason, when what I meant was `&quot;`. So to be clear, this is mudlet's current behavior: 

    >> say "Hello"
    << You say, &quot;Hello&quot;

I'll leave leave the original comments to make @BlauFeuer's reply make sense, but will add notes to them.

**Comment by Griatch** [comment]
I removed the `&quot;` replacement in the telnet MXP parser. This seems to resolve the problem for Mudlet while keeping it working in MUSHClient and in the webclient (the latter has not changed). So I hope this is now finally resolved and this issue is closed.

**Comment by vadi2** [comment]
Sorry about the dead IRC channel - we've moved development contact over to https://gitter.im/Mudlet/Mudlet. Will update the IRC topic.

Looks like this was resolved, though feel free to get in touch if you need anything!

**Comment by SlySven** [comment]
I'll also point you in the direction of our (Mudlet's) [user forums](http://forums.mudlet.org) and [bug tracker](https://bugs.launchpad.net/mudlet); sorry about the confusion about the old, but technically still the latest _release_ version 2.1 over on SourceForge - the only developer with write access has (_I think_) mislaid his credentials. As it happens we have been working from the GitHub repository that Vadim mentioned for several years now and we are getting close (Vadim hopes :stuck_out_tongue_winking_eye:) to _finally_ getting 3.0.0 out the door _real soon now™_.

As for MXP I am aware that Mudlet's current code has some shortcomings - e.g. it does not support the \<support> tag which is apparently a mandatory requirement for version 1.0 of that protocol :blush:, but as Vadim has tried to explain to me actually working out the exact protocol details does depend on who you ask about it :confused::hammer:.

**Comment by Griatch** [comment]
@vadi2, @SlySven Thanks for replying! We'll keep those new links in mind for future communications.
Yeah, supporting any protocol in MUD-land is a bit of a hit-and-miss. Because the definitions are occationally ... organic ... it's not not always clear exactly how something should work. And as server devs we find that even if a definition is seemingly clear, then different *clients* have interpreted it differently - so we have to support those *interpretations*. 
That said, I was not aware of the `<support>` tag being mandatory, so there you go. :smile:  Anyway, Mudlet (v3 version) seems to work fine for Evennia now (see also our [client support grid](https://github.com/evennia/evennia/wiki/Client-Support-Grid).

**Comment by whitehorse-io** [comment]
I'm still getting this issue with Mudlet 3.2.  Is there a way we can turn off MXP by default, so that Mudlet users connecting won't get the bug? It may not be an ideal work-around but Mudlet is a very popular mud client and we don't want to give anyone a bad impression when we launch, so to just turn off MXP by default in Evennia would work for now.

**Comment by BlauFeuer** [comment]
The problem is that it _is off by default._  Mudlet actually responds that it will respond to MXP, so Evennia turns it on.   Just post on your welcome screen about the issue with mudlet, and that the work-around is to set `@options/save MXP=off` until Mudlet fixes or user finds a new client.

**Comment by vadi2** [comment]
Where can this problem be reproduced?

**Comment by whitehorse-io** [comment]
Hi Vadi, we're making a MUD at 54.202.144.77:4000 and it's been reproducible for us using mudlet so far.

**Comment by vadi2** [comment]
Connecting to the game, I didn't see Evennia turning MXP on, though it says it will in the settings. Will have a look at this issue after sorting 3.3 out.

**Comment by BlauFeuer** [comment]
Ha, the extra line breaks are from the Telnet Go Ahead characters after the usual CR+LF.  Mudlet is treating the CR+LF + IAC + GA as terminating one line, then an empty string, then the IAC + GA is signaling a MUD prompt which gets displayed as another empty line.

So any transmission terminating with a Telnet Go Ahead will be displayed with what appears to be a second new line break.   If there's a checkbox for MUD prompt handling, try toggling it.  I think a prompt that's empty is an edge case that isn't being handled here.

**Comment by whitehorse-io** [comment]
Thanks Blau.  I don't understand this too well but is that a mudlet issue or something on my side with Evennia? I might have some old prompt-related code in there somewhere although the current MUD isn't displaying a prompt right now.

**Comment by BlauFeuer** [comment]
You can test it by trying `@option NOGOAHEAD=on` after you log in. I noticed the issue without logging in because the message for command not available are spaced out:
```
Command 'who' is not available. Type "help" for help.

Command 'what' is not available. Type "help" for help.
```

**Comment by BlauFeuer** [comment]
It's like you're using an older version of Mudlet or Evennia's options are going haywire on or after negotiation -- very strange, but very reproducible, so looking closely at the handshake and what happens afterward should reveal where the issue is,

**Comment by whitehorse-io** [comment]
You're right, that has fixed the line breaks! Hurray.  Is there a way I can make nogoahead True by default and is there a reason I would not want to do this?

**Comment by vadi2** [comment]
If Mudlet says <No GA> bottom right with that option enabled, then you
wouldn't want this option on - GA is needed by Mudlet to tell the ping time
and also for `isPrompt()` function to work. I'm wondering if the double GA
(as I understood @BlauFeuer) is correct - we've got a lot of other MUDs
using GA w/o this issue.

**Comment by BlauFeuer** [comment]
It's similar to the MXP negotiation, but for Go Ahead, it's reversed - Evennia defaults to on and the client has to either handle Go Ahead or ask for them to be suppressed.    

To work around it, set the option off after login.  I think isPrompt isn't dealing with empty string prompts properly - that's the edge case where a prompt of nothing is not a prompt at all.

Something's up with Mudlet and Evennia right now, not on the same page with these options and/or how they're interpreted and displayed.

**Comment by BlauFeuer** [comment]
My last thought on this is that perhaps Mudlet would respond differently if the end of transmission CR+LF + IAC + GA sequence was reversed to be IAC + GA+ CR+LF.  If that's how other mu* servers are sending the Go Ahead, then maybe Evennia should, also.   I'd have to connect to another mu* that does Telnet Go Ahead to understand why Mudlet is behaving for them, but not Evennia.

**Comment by Griatch** [comment]
Ok, reopening this again since the issues seem to have returned.

**Comment by SlySven** [comment]
BlauFeuer wrote:
> It's similar to the MXP negotiation, but for Go Ahead, it's reversed - Evennia defaults to on and the client has to either handle Go Ahead or ask for them to be suppressed.

"Evennia defaults to on" hang on - that sounds suspect! All Telnet sub-options (including Go Ahead) are supposed to be __disabled__ until they get negotiated see [RFC854](https://tools.ietf.org/html/rfc854) and successors {mind you I am not totally sure Mudlet is correct for _every_ option either...}

**Comment by SlySven** [comment]
Should a MUD Server wish to identify it is talking to a _Mudlet_ Client the latter does respond to Telnet SubOption 24 {TTYPE} and will return a string (which according to the specification for that is case insensitive and a maximum of 40 bytes - though I have not put a clamp on that (yet) in Mudlet) of the form "Mudlet 3.x.y" (with a possible "-buildDesignation" if it is __NOT__ a release version) where 3.x.y is the semantic version number.  We do not yet support [MTTP](http://tintin.sourceforge.net/mtts/) although that is on my TODO _long_-list.

**Comment by BlauFeuer** [comment]
http://freesoft.org/CIE/RFC/1123/23.htm regarding RFC854 and 858 is where I read "Telnet Server MUST attempt to negotiate the Suppress Go Ahead option (i.e., send "WILL Suppress Go Ahead"). A User or Server Telnet MUST always accept negotiation of the Suppress Go Ahead option. "  meaning it's ON after handshaking unless client asks for suppression, not to imply that it's ON before handshaking.  

But the point here is that Mudlet handles Go Ahead when there's no MUD prompt being sent by displaying a blank prompt which shows as an extra new line.  That's the behavior in question regarding interpreting CR+LF + IAC+GA.

And then there's the on-topic issue of Mudlet reporting that it wants MXP replacement at the handshaking, and the server then must send all literal `< >` as their `&lt;, &gt;` html counterparts, since `<>` tags are reserved for use in MXP mode.

That's about the limit of my understanding of the two issues here.

**Comment by SlySven** [comment]
... and to be honest some of the details are not clear to me either... 😝

**Comment by SlySven** [comment]
> ...  A User or Server Telnet MUST always accept negotiation of the Suppress Go Ahead option. " meaning it's ON after handshaking unless client asks for suppression, not to imply that it's ON before handshaking.

Yeah it must support negotiating about the option (it receives a `WILL GA` - but it has the right to decline it `DONT GA`).

just looking at our code-base has shown a possible oddity with a pair of _LineFeed on Go-Ahead_ flags but I'd need to check that on a clean code-base (the code I am currently working with won't compile until I finish the bit I am currently frobbing...)

**Comment by Griatch** [comment]
@SlySven: Any update on this in respect to the possible oddity you mention? Best to ask before we start to look into our code on this end. :)

**Comment by Griatch** [comment]
@TehomCD I cannot reproduce this with a vanilla install (fresh install, no previous settings) of [latest Mudlet 3.4.0 under Linux](https://www.mudlet.org/download/#). I can don't see > turn into its html-escaped counterpart regardless of if `MXP` is on or off. Could someone else verify?

**Comment by chainsol** [comment]
I can confirm - latest Mudlet seems to have no problems, even in raw mode, or with mxp off.

**Comment by drakmirv** [comment]
I can confirm that latest Mudlet 3.5.0 on Windows shows this same behavior (&lt; and &gt;).  Turning off MXP either in Mudlet or Evennia shows the proper characters.

Thus, I think there is still an issue with Mudlet and Evennia when MXP mode is on.

**Comment by drakmirv** [comment]
More information.  I can click "reconnect" multiple times in Mudlet and have this behavior "toggle" between showing the xml attributes and not.  When the attributes are not showing on the connection screen, then all 'seems' ok and @options reports that MXP mode is on.  However, I don't think it really is.

Toggling @options mxp=false seems to make no change, but @options mxp=true brings back the error.

Therefore, I think there is some problems with protocol negotiation between Mudlet and the Evennia server.

**Comment by Griatch** [comment]
@drakmirv Interestingly, this appears to vary from Mudlet version to Mudlet version or possibly from OS to OS. If you revert to Mudlet 3.4, do you see the same behavior?

**Comment by drakmirv** [comment]
Same behavior on 3.4.0 and 3.6.0.

**Comment by SlySven** [comment]
I am not aware of any recent changes in the area of Mudlet's handling of the MXP option!

Can I check does Evennia send a IAC WILL OPT_MXP to request that it uses MXP to take to Mudlet (which will be ignored if the "Preferences" option in Mudlet to ignore MXP is ☑️ {checked}) - otherwise I can see Mudlet will return the IAC DO OPT_MXP to tell Evennia to use it.

Looking a bit deeper in the code involved I can see some flaws in the Mudlet code (in `cTelnet::processTelnetCommand(const string&)` in the `src/cTelnet.cpp`) which means it does not update it's internal state variables (in five 256 `bool` arrays) when some options that it explicitly does handle are processed - the execution flow `break`s out of some `switch(option)` statements before it gets to those state variables, and those same state variables are largely ignored anyway in favour of a seeminglying ad-hoc collection of individual `bool` variables whose direction of effect for those that have that function (is it for the Server to Mudlet or Mudlet to Server - or BOTH directions) is not obvious... 🤓 

Trouble is that that particular part of the code base is a cross between Topsy and Frankenstein's monster and the fact that it _sort of_ works means there has to be much loin girding before anyone does anything to it... 😮

**Comment by Griatch** [comment]
@SlySven A part of Evennia's handshake is `IAC WILL_MXP`, to which it expects the client to respond with `IAC DO_MXP`. One issue might be if the client takes too long to respond maybe - Evennia's Portal will grow bored of waiting after a while and go with defaults when syncing the status to the Server; maybe a late response from the client is not correctly passed on...

**Comment by Griatch** [comment]
@SlySven, @vadi2  So I've examined this more in detail. I have on my local machine fixed a few edge cases that made evennia's MXP setting less reliable than it could be (basically it got over-ridden by the GMCP message with client name Mudlet sent, this was all on us). However, it appears that Mudlet makes some very strange assumptions about the use of < and > coming from the server.

When Evennia connects it sends a greeting screen, which includes the < > and symbols (these appear e.g. in the text `connect <username> <password>`). At this point no hand shake has had time to finish so Evennia will assume no more than plain ANSI colors. But mudlet will promptly ignore < > completely, removing them and the text inside them from the display (you'll only see `connect ` with nothing following it). 
Note - no MXP *can* be active at this point since the handshakes have not had time to complete. Evennia will then send `WILL MXP` and Mudlet will reply `DO MXP`. At this point Evennia turns on MXP and escapes < > into its HTML counterparts. So from here on, once you do `look` to re-display the greeting screen, the <> will now appear correctly.

 One could imagine that Mudlet returns `WILL MXP` and turns on its MXP parsing before Evennia gets the reply - but this doesn't seem to be the case: If I deliberately make sure to delay the `DO MXP` call to happen well after that first greeting screen, Mudlet will still fail to parse <> (and the time of failed display just becomes longer). Not until Evennia has correctly activated MXP will Mudlet correctly display < and  >, 

So while some of the inconsistencies in testing this has been on Evennia (I've not yet pushed the changes to those), it appears Mudlet somehow assumes escaped < and > even though the server has not yet acknowledged that it will escape according to the MXP standard.

**Comment by Griatch** [comment]
I implemented a slight 0.3s delay in the first send of the greeting screen. That is enough to allow mudlet to respond and the MXP to set up correctly. This solution is network-dependent though so it will not necessarily work for poor networks. I think this needs to be fixed at the Mudlet side. Closing this for now, re-open if further issues arise.

**Comment by Griatch** [comment]
@SlySven, @vadi2 Opening this again, because after a reboot of mudlet it now fails consistently again (I don't know why the same version can behave so inconsistently. A timing issue, different loads on my computer?). This is the sequence as far as I interpret it: 

0. Connection is established. 
1. Evennia sends telnet subnegotiation `DO MXP `
2. At this point Evennia doesn't use MXP so it actually (now) delays the welcome screen for 0.3 seconds to give the client a chance to respond with eventual special abilities to show on the welcome screen  (but this doesn't appear to matter for Mudlet)
3. Mudlet returns subnegotiation `WILL MXP`. Evennia now turns on MXP parsing, which means <> will be translated to their HTML equivalents as per the MXP specification.
4. Usually the welcome screen is now sent from Evennia since 0.3s has passed. If it is sent before Mudlet has returned with `WILL MXP`, it will be sent with unparsed <> which looks fine in Mudlet. In the end it doesn't matter though, once Evennia is told to switch the MXP, mudlet displays the tags verbatim. 
5. Mudlet displays the HTML tags verbatim and will continue to do so. Only way to fix this is to manually turn off MXP in Evennia (or to switch Mudlet to no-MXP mode). 

Overall, the perception here is that Mudlet tells Evennia to turn on MXP but then doesn't actually do so itself. Is Mudlet expecting yet another confirmation from Evennia somehow?

**Comment by drakmirv** [comment]
I think this is backwards.  That sequence is telling the client that the server wants to receive MXP.

Please see: 
http://users.cs.cf.ac.uk/Dave.Marshall/Internet/node141.html
http://mud-dev.wikidot.com/telnet:negotiation
http://www.gammon.com.au/forum/bbshowpost.php?bbsubject_id=221

Evennia should send:
IAC WILL MXP

Client sends
IAC DO MXP   -or- IAC DONT MXP

**Comment by Griatch** [comment]
@drakmirv Yeah, I mixed up DO and WILL when writing the list. As you say, it's the other way around and that's also what Evennia does. I corrected the previous comment, thanks.

**Comment by Griatch** [comment]
And I've actually not seen that 3rd subnegotiation option before, the official MXP docs at http://www.zuggsoft.com/zmud/mxp.htm

**Comment by vadi2** [comment]
Thanks for doing this research - we're a bit tied up right now but
hopefully someone will have a look at this soon.

**Comment by Griatch** [comment]
I'm not sure what more I can do with this issue at this point, so marking it as `help wanted` in case someone else has any more ideas.

**Comment by bornasm** [comment]
I played around with this issue today and I have some potentially useful info.

Mudlet doesn't show < and > properly until the first time I use |lc|lt|le somewhere.

Temporary workaround for the issue I currently use is just placing "|lc|lt|le" anywhere in the connection screen. It doesn't render anything but fixes the issue.

**Comment by SlySven** [comment]
Um, what sort of magic incantation *is* `|lc|lt|le`?

**Comment by bornasm** [comment]
https://github.com/evennia/evennia/wiki/TextTags#clickable-links

**Comment by vadi2** [comment]
I'm looking at this now - and forgive me it's been ages since I worked with MXP - but from the looks of it Evennia never sends an MXP control tag when it escapes `<` and `>` on the connection screen in `demo.evennia.com`. Mudlet checks for codes `1`, `4` and `6` for these. @Griatch could you confirm?

**Comment by Griatch** [comment]
@vadi2 I'll need to look at this again, but from the MXP definition, there should be no further MXP control tag to escape `<` and `>` once MXP is activated. 
The way I understand it: 
- When server/client has not yet agreed to send MXP - send `<` and `>` as-is
- When both have agreed to activate MXP, use raw `<` and `>` only for MXP-related tags (in Evennia's case this is only for clickable links), while using `&lt;` and `&gt;` whenever `<` or `>` is to be sent on their own. Whereas the MXP specs are not the best writing from a standards perspective, this appears reasonable and also how clients like MUSHClient appear to handle it (It's clarified in their forums as I recall). 

This is interesting though, what do you mean by Mudlet expecting "codes 1, 4 or 6"? Could you give an example of how you expect Evennia to send its `<` and `>` pre/post MXP negotiation?

**Comment by vadi2** [comment]
I think you nailed it - and this is a suspicion I had when I was looking at the MXP docs for this last week. Our current code only deserialises it's seen at least an MXP tag, and we've just never ran into a case where the comparison operator is sent before a tag consistently.

We'll look into fixing it :)

**Comment by vadi2** [comment]
@Griatch Looking at it again. Can you point me to where `<` `>` should always be escaped? Doesn't:

>    no MXP or HTML commands are allowed in the line.  The line is not parsed for any tags at all.  This is useful for "verbatim" text output from the MUD.  When a newline is received from the MUD, the mode reverts back to the Default mode.

Imply that they shouldn't be?

**Comment by vadi2** [comment]
Fixed in development, here are test builds:

https://ci.appveyor.com/api/buildjobs/696de146ac3hermo/artifacts/src%2Fmudlet.zip (windows)
https://transfer.sh/rRLXe/Mudlet-3.16.1-testing-PR2198-48ca5bf.dmg (macOS)
https://transfer.sh/Eb0sI/Mudlet-3.16.1-testing-PR2198-48ca5bf-linux-x64.AppImage.tar (linux)

**Comment by SlySven** [comment]
That temporary secure mode (4) only lasts for the next *\<tag>* (not the next MXP *command*) so the SGR code for it is needed again for the corresponding *\</tag>* - as the specification points out...

Though *IMVHO* I think we (Mudlet) still need to work on this.

**Comment by Griatch** [comment]
@SlySven : Correct, I was sloppy with the nomenclature and should have said "tag" rather than "command". The way we do it is indeed thus: 

1. Let's say the incoming string is `"Please |lc look |lt <look here>! |le"`, where `|lc`, `|lt` and `|le` are Evennia's own syntax for adding a link in order to, in this case, fire the "look" command. 
2. Since there are <> in there, these are now escaped to `"Please |lc look |lt &lt;look here&gt;! |le"`
3. Next, it is converted to the MXP structure: 
    `"Please  MXP_TEMPSECURE  <SEND HREF="look"> &lt;look here&gt;! MXP_TEMPSECURE </SEND>`

Note that the original text must be escaped, not only because it's outside of the `MXP_TEMPSECURE` mode (which ends with each tag), but also logically, since otherwise those <> would mess with the MXP command. 

... But that said, the issue with Evennia and Mudlet is not inside these MXP links but outside of them it would appear. I first thought it could be that Mudlet is not correctly closing that temp-secure MXP mode with that last `MXP_TEMPSECURE </SEND>`, but if so I'd expect things to stop working after the link. That's not the case - it's confusing since people report that things will start to *work* once you have an MXP link in there ...

**Comment by Griatch** [comment]
Mudlet 3.18+ no longer exhibits this behavior (fixes applied on the Mudlet side, it appears). Closing (again).
